File: bench/diff_strncpy_from_user.c
// SPDX-License-Identifier: GPL-2.0
// Tier-2.5 differential oracle: C original vs Rust translation,
// strncpy_from_user. Isolates the fault-free word-at-a-time
// scanning/copying arithmetic from lib/strncpy_from_user.c -- the real
// unsafe_get_user() fault path can't be triggered portably from a host
// program (no way to cause a real page fault at a chosen address and
// resume via extable outside the kernel), so this oracle only exercises
// do_strncpy_from_user()'s arithmetic with `max` always large enough
// that no fault occurs; the fault path itself is inspection-verified +
// boot-verified only, same as strnlen_user's landing (see this file's
// sibling in the same commit, and lib/strnlen_user_rs.rs's module doc
// for the identical constraint).
//
// Faithfully reproduces both loops (word-at-a-time + byte-at-a-time
// fallback) and the IS_UNALIGNED gate exactly as
// lib/strncpy_from_user.c defines them for this build
// (CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS unset -- see
// lib/strncpy_from_user_rs.rs's module doc), so alignment-driven
// fallback is genuinely exercised, not skipped.
//
// do_strncpy_from_user()'s -EFAULT return has TWO distinct triggers in
// the real kernel: (a) a genuine unsafe_get_user() page fault (the
// `goto efault`/`goto byte_at_a_time` targets) -- NOT reproducible on
// a host program, inspection- and boot-verified only; and (b) the
// pure-arithmetic "hit max, but max < count" post-loop check
// (`if (res >= count) return res; ... return -EFAULT;`), which needs
// no real fault at all -- just max < count, driven entirely by the
// address-space-limit truncation the caller (strncpy_from_user) does
// before calling in. This oracle deliberately drives max < count
// (mode 5 below) to exercise (b) for real, in addition to the
// count<=0 / truncation / exact-fit / full-copy classes.
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef unsigned long ul;
#define EFAULT 14

#define IS_UNALIGNED(src, dst) (((long)(dst) | (long)(src)) & (sizeof(long) - 1))

#define ONE_BITS (~0ul / 0xff)
#define HIGH_BITS (ONE_BITS * 0x80)
static inline ul has_zero(ul val, ul *bits)
{
	ul mask = ((val - ONE_BITS) & ~val) & HIGH_BITS;
	*bits = mask;
	return mask;
}
static inline ul create_zero_mask(ul bits)
{
	bits = (bits - 1) & ~bits;
	return bits >> 7;
}
static inline int my_fls64(uint64_t x)
{
	if (x == 0) return 0;
	return 64 - __builtin_clzll(x);
}
static inline ul find_zero(ul mask) { return my_fls64(mask) >> 3; }
static inline ul zero_bytemask(ul mask) { return mask; } // riscv identity

// do_strncpy_from_user, byte-replicated from lib/strncpy_from_user.c.
// `src`/`dst` are plain host pointers here (no real userspace fault
// boundary); `max` is always chosen by the driver loop below to be
// >= the generated string's length + 1, so the byte_at_a_time loop's
// `max` exhaustion path and the final -EFAULT branch are exercised via
// the COUNT-vs-length boundary sweep (max==count in all cases below),
// never via a genuine out-of-bounds read.
static long my_do_strncpy_from_user(char *dst, const char *src, unsigned long count, unsigned long max)
{
	unsigned long res = 0;

	if (IS_UNALIGNED(src, dst))
		goto byte_at_a_time;

	while (max >= sizeof(ul)) {
		ul c, data, mask, bits;

		memcpy(&c, src + res, sizeof(ul));

		if (has_zero(c, &bits)) {
			data = create_zero_mask(bits);
			mask = zero_bytemask(data);
			ul masked = c & mask;
			memcpy(dst + res, &masked, sizeof(ul));
			return res + find_zero(data);
		}

		memcpy(dst + res, &c, sizeof(ul));

		res += sizeof(ul);
		max -= sizeof(ul);
	}

byte_at_a_time:
	while (max) {
		char c = src[res];
		dst[res] = c;
		if (!c)
			return res;
		res++;
		max--;
	}

	if (res >= count)
		return res;

	return -EFAULT;
}

// strncpy_from_user, minus might_fault/should_fail_usercopy/
// kasan_check_write/check_object_size (all no-ops for this build, see
// lib/strncpy_from_user_rs.rs's module doc) and minus the
// TASK_SIZE_MAX/untagged_addr address-space-limit dance (no real
// userspace address space on the host -- max is passed in directly by
// the test driver, exactly mirroring "src_addr < max_addr" always
// holding and max already truncated to count, which is the common
// case this oracle targets).
static long my_strncpy_from_user(char *dst, const char *src, long count, unsigned long max)
{
	if (count <= 0)
		return 0;
	return my_do_strncpy_from_user(dst, src, (unsigned long)count, max);
}

// Explicit LCG (same constants as bench/diff_base64.c / diff_string.c).
static uint64_t lcg_state;
static uint32_t lcg_next(void)
{
	lcg_state = lcg_state * 6364136223846793005ULL + 1442695040888963407ULL;
	return (uint32_t)(lcg_state >> 32);
}

static const char ALPHABET[] = "abcXYZ .";
static void gen_str(char *buf, int maxlen)
{
	int len = 1 + lcg_next() % (maxlen - 1);
	for (int i = 0; i < len; i++) buf[i] = ALPHABET[lcg_next() % (sizeof(ALPHABET) - 1)];
	buf[len] = 0;
}

#define BUFLEN 96

int main(int argc, char **argv)
{
	long n = argc > 1 ? atol(argv[1]) : 5000;
	lcg_state = argc > 2 ? (uint64_t)atol(argv[2]) : 424242;

	// Backing arrays deliberately oversized vs BUFLEN and offset by a
	// runtime-chosen byte so src/dst alignment varies across cases --
	// exercises IS_UNALIGNED's real branch, not just the aligned arm.
	char src_backing[BUFLEN + 16];
	char dst_backing[BUFLEN + 16];

	for (long i = 0; i < n; i++) {
		int src_off = lcg_next() % 8;
		int dst_off = lcg_next() % 8;
		char *src = src_backing + src_off;
		char *dst = dst_backing + dst_off;

		memset(src_backing, 0, sizeof(src_backing));
		memset(dst_backing, 0x55, sizeof(dst_backing));

		gen_str(src, BUFLEN - 16);
		size_t slen = strlen(src);

		// count sweeps across, below, at, and above slen+1 (the NUL
		// position) -- the exact boundary sweep the task calls for:
		// alignment x NUL-position x count-boundary.
		long count;
		int mode = lcg_next() % 6;
		switch (mode) {
		case 0: count = 0; break;                              // count<=0 edge
		case 1: count = -(long)(lcg_next() % 4) - 1; break;     // negative count
		case 2: count = (long)(lcg_next() % (slen + 1)); break; // < strlen+1: truncation
		case 3: count = (long)(slen + 1); break;                // exact boundary
		case 4: count = (long)(slen + 1 + lcg_next() % 8); break; // > strlen+1
		default: count = (long)(slen + 1 + lcg_next() % 8); break; // mode 5: max<count below
		}

		// max == count for every mode except 5, which deliberately
		// sets max < count (a real "hit the address-space budget
		// before satisfying count" case -- pure arithmetic, see file
		// header point (b)) to exercise the non-fault -EFAULT return.
		unsigned long max = count > 0 ? (unsigned long)count : 0;
		if (mode == 5 && max > 0)
			max = lcg_next() % max; // strictly less than count

		long r = my_strncpy_from_user(dst, src, count, max);

		printf("strncpy_from_user,%s,%ld,%ld,[%.*s]\n",
		       src, count, r, r > 0 ? (int)r : 0, r > 0 ? dst : "");
	}
	return 0;
}

File: bench/diff_strncpy_from_user.rs
// SPDX-License-Identifier: GPL-2.0
//! Tier-2.5 differential oracle: Rust translation side,
//! strncpy_from_user. Faithful host-side copy of
//! lib/strncpy_from_user_rs.rs's do_strncpy_from_user algorithm
//! (kernel-crate bindings/export/unsafe_get_user shims stripped,
//! replaced with plain slice reads since there is no real userspace
//! fault boundary on the host) -- same protocol/LCG and same
//! fault-free-arithmetic-only scope as diff_strncpy_from_user.c (see
//! that file's header comment for the two distinct -EFAULT triggers
//! and why mode 5's max<count case genuinely exercises one of them).

fn has_zero(val: usize, one_bits: usize, high_bits: usize) -> usize {
    ((val.wrapping_sub(one_bits)) & !val) & high_bits
}
fn create_zero_mask(bits: usize) -> usize {
    let bits = bits.wrapping_sub(1) & !bits;
    bits >> 7
}
fn fls64(x: u64) -> i32 {
    if x == 0 {
        0
    } else {
        (63 - x.leading_zeros() as i32) + 1
    }
}
fn find_zero(mask: usize) -> usize {
    (fls64(mask as u64) >> 3) as usize
}

fn is_unaligned(src: *const u8, dst: *const u8) -> bool {
    ((dst as isize) | (src as isize)) & (core::mem::size_of::<usize>() as isize - 1) != 0
}

// Mirrors lib/strncpy_from_user_rs.rs's do_strncpy_from_user exactly
// (same two-loop shape, same word/byte boundary arithmetic), minus the
// unsafe_get_user fault-shim calls (plain slice reads instead -- no
// fault possible on a host Vec, and this oracle never lets `max`
// exceed the backing buffer, see file header) and the SAFETY comments
// (no unsafe user-pointer contract on the host side).
fn do_strncpy_from_user(dst: &mut [u8], src: &[u8], count: usize, max: usize) -> isize {
    let mut res: usize = 0;
    let mut max = max;

    let src_ptr = src.as_ptr();
    let dst_ptr = dst.as_ptr();

    if !is_unaligned(src_ptr, dst_ptr) {
        while max >= core::mem::size_of::<usize>() {
            let c = usize::from_ne_bytes(
                src[res..res + core::mem::size_of::<usize>()]
                    .try_into()
                    .unwrap(),
            );

            const ONE_BITS: usize = usize::MAX / 0xff;
            const HIGH_BITS: usize = ONE_BITS * 0x80;
            let data = has_zero(c, ONE_BITS, HIGH_BITS);
            if data != 0 {
                let data = create_zero_mask(data);
                let mask = data; // zero_bytemask(mask) == mask on riscv
                let masked = c & mask;
                dst[res..res + core::mem::size_of::<usize>()]
                    .copy_from_slice(&masked.to_ne_bytes());
                return (res + find_zero(data)) as isize;
            }

            dst[res..res + core::mem::size_of::<usize>()].copy_from_slice(&c.to_ne_bytes());

            res += core::mem::size_of::<usize>();
            max -= core::mem::size_of::<usize>();
        }
    }

    // byte_at_a_time:
    while max != 0 {
        let c = src[res];
        dst[res] = c;
        if c == 0 {
            return res as isize;
        }
        res += 1;
        max -= 1;
    }

    if res >= count {
        return res as isize;
    }

    -14 // -EFAULT
}

fn strncpy_from_user(dst: &mut [u8], src: &[u8], count: isize, max: usize) -> isize {
    if count <= 0 {
        return 0;
    }
    do_strncpy_from_user(dst, src, count as usize, max)
}

// Identical LCG to diff_strncpy_from_user.c.
struct Lcg(u64);
impl Lcg {
    fn next(&mut self) -> u32 {
        self.0 = self
            .0
            .wrapping_mul(6364136223846793005)
            .wrapping_add(1442695040888963407);
        (self.0 >> 32) as u32
    }
}

const ALPHABET: &[u8] = b"abcXYZ .";

fn gen_str(rng: &mut Lcg, maxlen: usize) -> Vec<u8> {
    let len = 1 + (rng.next() as usize) % (maxlen - 1);
    let mut v = Vec::with_capacity(len + 1);
    for _ in 0..len {
        v.push(ALPHABET[(rng.next() as usize) % ALPHABET.len()]);
    }
    v.push(0);
    v
}

const BUFLEN: usize = 96;

fn cstr(v: &[u8]) -> &str {
    let end = v.iter().position(|&b| b == 0).unwrap_or(v.len());
    std::str::from_utf8(&v[..end]).unwrap()
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let n: u64 = args.get(1).and_then(|a| a.parse().ok()).unwrap_or(5000);
    let seed: u64 = args.get(2).and_then(|a| a.parse().ok()).unwrap_or(424242);
    let mut rng = Lcg(seed);

    for _ in 0..n {
        // Backing arrays oversized vs BUFLEN and offset by a
        // runtime-chosen byte so src/dst alignment varies across
        // cases -- exercises IS_UNALIGNED's real branch, matching
        // diff_strncpy_from_user.c's approach exactly.
        let src_off = (rng.next() as usize) % 8;
        let dst_off = (rng.next() as usize) % 8;

        let mut src_backing = vec![0u8; BUFLEN + 16];
        let mut dst_backing = vec![0x55u8; BUFLEN + 16];

        let s = gen_str(&mut rng, BUFLEN - 16);
        src_backing[src_off..src_off + s.len()].copy_from_slice(&s);
        let slen = s.len() - 1; // gen_str includes trailing NUL in its Vec

        let mode = rng.next() % 6;
        let count: isize = match mode {
            0 => 0,
            1 => -((rng.next() % 4) as isize) - 1,
            2 => (rng.next() as usize % (slen + 1)) as isize,
            3 => (slen + 1) as isize,
            4 => (slen + 1 + (rng.next() as usize % 8)) as isize,
            _ => (slen + 1 + (rng.next() as usize % 8)) as isize, // mode 5: max<count below
        };

        // max == count for every mode except 5, which deliberately
        // sets max < count (a real "hit the address-space budget
        // before satisfying count" case -- pure arithmetic, see
        // diff_strncpy_from_user.c's file header point (b)) to
        // exercise the non-fault -EFAULT return.
        let mut max: usize = if count > 0 { count as usize } else { 0 };
        if mode == 5 && max > 0 {
            max = (rng.next() as usize) % max; // strictly less than count
        }

        let src = &src_backing[src_off..];
        let dst = &mut dst_backing[dst_off..];

        let r = strncpy_from_user(dst, src, count, max);

        let out = if r > 0 { cstr(&dst[..r as usize]) } else { "" };
        println!("strncpy_from_user,{},{},{},[{}]", cstr(&s), count, r, out);
    }
}
